Publish events to Spring Cloud Stream channels

mehdichitforoosh opened the pull request:
We can integrate Axon `EventBus` with Spring Integration and Spring Cloud Stream with these configurations.
As a Message Producer:
```
    @Bean
    public IntegrationFlow flow(EventBus eventBus) {
        return IntegrationFlows.from(new AxonInboundChannelAdapter(eventBus))
                .channel(Source.OUTPUT)
                .get();
    }
```
As a Consumer:
  ```
   @Bean
    public IntegrationFlow flow(EventBus eventBus) {
        return IntegrationFlows.from(Sink.INPUT)
                .handle(new AxonOutboundChannelAdapter(eventBus))
                .get();
    }

    @Autowired
    public void configure(EventProcessingConfigurer config) {
        config.usingSubscribingEventProcessors();
    }
```

`Sink.INPUT` and `Source.OUTPUT` are `MessageChannel`s from Spring Cloud Stream.

smcvb commented:
@mehdichitforoosh, are you up for adjusting your PR towards a different format?
As it stands, I think there's value in having an option to connect to Spring Cloud Streams for events.
However, I don't think this should be contained inside the module you've dropped it in right now.

In essence, this extension has been set up to tap into Spring Cloud Discovery for Command Message distribution.
What you are providing is a solution to tap into Spring Cloud Stream for Event Message distribution.

That would thus opt for the current code to be moved to an `axon-springcloud-discovery` module, whereas your code would reside in an `axon-springcloud-streams` module.
If you want to take it to that end, please let us know.
In absence of response, which might be imaginable after such a long timeslot, we'll likely pick up your code and do this ourselves eventually.

mehdichitforoosh commented:
Hi @smcvb.Yes,I want to complete this PR.
What changes should I make?

smcvb commented:
That's amazing to hear @mehdichitforoosh!
As first stabs to revamp this PR, I'd like you to do the following:

1. Add a distinct module to this project, dedicated to this solution. Calling the package `springcloud-streams` and the artifact `axon-springcloud-streams` would clarify this is different from the existing Spring Cloud Discovery implementation.
2. Repoint this PR towards the most recent version of the extension. This might be some effort, true, but it would be most reasonable to think of releasing this with a recent version of the extension. Pointing your PR to `master` and merging it with the current changes in `master` should be sufficient to achieve this.

mehdichitforoosh commented:
Ok @smcvb 
Well, I will definitely do it and I will inform you in the next few days.
Thanks.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/AxonFramework/extension-springcloud?pullRequest=5) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/AxonFramework/extension-springcloud?pullRequest=5) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: smcvb<br/>:x: mehdichitforoosh<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/AxonFramework/extension-springcloud?pullRequest=5) it.</sub>

mehdichitforoosh commented:
dear @smcvb 
I added changes.
What should I do for the next step?

smcvb commented:
Thanks for the effort so far @mehdichitforoosh, much appreciated.
For now, I'd like to ask you to do the following things:

- Add descriptive class-level Javadoc to any of the classes you have added. When doing so, don't forget to adjust the `@author` (with your preferred name, of course) and `@since` (pointing towards 4.5, for now) tags correctly.
- Using "...Axon..." in any of the class names shouldn't be necessary I think. As long as the implementation describes the intent of that class in the translation from Spring-Streams to Axon's idea of Events, Event Processors and Message Sources, we should be good. The `AxonProcessorMessageHandler` to me for example sounds like a `SpringStreamsMessageSource`.
- Any of the copyright notices should stretch towards 2021.
- A lot of the classes seem to be duplicated right now. Make sure only a single instance of each is present.
- I am missing unit test classes for the provided code. We definitely need these prior to approve this PR.
- For components like these, a short sample implementation through an integration test would be highly beneficial. This would ensure the process works as desired, but can also serve as a sample to others who want to use this extension.

Likely there are more fine-grained comments to go over too, but let's first take a stab at the above.
Again, thanks for the effort so far @mehdichitforoosh! 
If your time doesn't allow to proceed with this, I think there might be an option that we'll (at AxonIQ) take it over.
Just let us know whether you prefer that.

mehdichitforoosh commented:
Thanks @smcvb  for your attention.
I will finish this PR soon :-)

mehdichitforoosh commented:
Hi Steven @smcvb 
I think this PR is completed.
1. I added Javadoc 
2. Rename and refactor all classes
3. Add copyright texts
4. Removed duplicate classes
5. Unit classes with Junit5, Mockito and Spring cloud stream test support
Thanks for your patience.
Tell me if there is a problem in latest commits.
Good luck.

mehdichitforoosh commented on springcloud/src/main/java/org/axonframework/extensions/springcloud/eventhandling/processor/AxonProcessorMessageHandler.java:
```diff
@@ -0,0 +1,110 @@
+     * @param filter                The filter that indicates which messages to forward.
+     * @param eventMessageConverter The message converter to use to convert spring message into event message
+     */
+    public AxonProcessorMessageHandler(SubscribableMessageSource<EventMessage<?>> messageSource, List<Consumer<List<? extends EventMessage<?>>>> processors
```
@idugalic Thanks for your comments dear  Ivan.

mehdichitforoosh commented:
Hi @smcvb.
Did you check commits?
Tell me if there is a problem with the code.
Thank you.

idugalic commented on springcloud-streams/src/main/java/org/axonframework/extensions/springcloud/eventhandling/SpringStreamMessageProcessor.java:
```diff
@@ -0,0 +1,207 @@
+        this.messageSource = builder.messageSource;
+        this.converter = builder.converter;
+        if (builder.eventBus != null) {
+            this.messageProcessors.addAll(singletonList(builder.eventBus::publish));
```
I wonder if we should add EventBus by default here? @smcvb what do you think? In my opinion, we could leave this decision to the users of the API, but I am afraid that in that case, we are missing something like `MultiSubscribableMessageSource.java` allowing them to subscribe to multiple sources: eventBus/eventStore, streamSource, ...

idugalic commented:
Hi @mehdichitforoosh. Thank you for this PR. Much appreciated! Conceptually this looks very good! 

It is important to note that this extension module is supporting `SubscribableMessageSource` only. The subscribable stream leaves all the ordering specifics in the hands of brokers (Kafka, for example), which means the events should be published on a consistent partition to ensure ordering. **I am totally fine with this, and this is not a blocker or stopper in my opinion**. To control events of a certain group to be placed in a dedicated partition, based on an aggregate identifier, for example, the message converter's SequencingPolicy can be utilized.` StreamableMessageSource` could be considered, but it is not something I would push now (let's let the brokers do the job).

mehdichitforoosh commented:
Hi @idugalic Thank you for your comments.

smcvb commented on pom.xml:
```diff
@@ -52,8 +53,11 @@
+        <spring-cloud-stream.version>Horsham.SR11</spring-cloud-stream.version>
 
         <spring.version>5.3.4</spring.version>
+        <spring-messaging.version>5.3.4</spring-messaging.version>
```
Is there an actual need to have a separate `spring-messaging` and `spring-integration` property, or could we tag along with `spring.version`?

smcvb commented on springcloud-streams/src/main/java/org/axonframework/extensions/springcloud/eventhandling/SpringStreamMessageSource.java:
```diff
@@ -0,0 +1,150 @@
+ */
+public class SpringStreamMessageSource extends AbstractMessageProducingHandler implements SubscribableMessageSource<EventMessage<?>> {
+
+    private final CopyOnWriteArrayList<Consumer<List<? extends EventMessage<?>>>> messageProcessors = new CopyOnWriteArrayList<>();
```
I think it's better to call these `eventProcessors` as they are consumers of the `EventMessage` type.

smcvb commented on springcloud-streams/src/main/java/org/axonframework/extensions/springcloud/eventhandling/SpringStreamMessageSource.java:
```diff
@@ -0,0 +1,150 @@
+     */
+    @Override
+    protected void handleMessageInternal(Message<?> message) {
+        Optional<EventMessage<?>> optional = converter.readSpringMessage(message);
+        optional.ifPresent(eventMessage -> {
```
Nit: Why not chain the outcome of the `converter` directly with the `ifPresent` invocation?

smcvb commented on springcloud-streams/src/main/java/org/axonframework/extensions/springcloud/eventhandling/SpringStreamMessageSource.java:
```diff
@@ -0,0 +1,150 @@
+         * @param eventBus The messageSource to subscribe
+         * @return the current Builder instance, for fluent interfacing
+         */
+        public SpringStreamMessageSource.Builder eventBus(EventBus eventBus) {
```
I think we can remove the `EventBus` from the builder, as described earlier in the constructor.

smcvb commented on springcloud-streams/src/main/java/org/axonframework/extensions/springcloud/eventhandling/SpringStreamMessageSource.java:
```diff
@@ -0,0 +1,150 @@
+         * @param converter The converter to convert messages
+         * @return the current Builder instance, for fluent interfacing
+         */
+        public SpringStreamMessageSource.Builder converter(SpringMessageConverter converter) {
```
As we're in the `Builder` here, we do not have to define the root class:

```suggestion
        public Builder converter(SpringMessageConverter converter) {
```

mehdichitforoosh replied:
@smcvb  No Problem.
Thank you for your reviews.
I'll check your comments and commit again in a few days.
Good luck.

smcvb replied:
Sure thing @mehdichitforoosh! I'll aim to get you an earlier follow-up review this time around :+1:

smcvb reviewed (changes requested):
For now, I will not proceed with the integration tests yet. 
Got a bunch of comments here, lots about JavaDoc. My main concerns are the usage of the `EventBus` in the message source directly and the seeming duplication with the `SpringStreamMessageProcessor` and `SpringStreamMessageSource

And, I shouldn't forget to extend my apologies for the late review.
Times have been hectic here @mehdichiforoosh, and sadly there weren't a lot of people we could spare on the other subjects we're working on.
I hope this review doesn't come too late for you..`

smcvb commented:
Closing this PR in absence of recent replies. If your time allows it, @mehdichitforoosh, be sure to reach out again; I'll gladly reopen it at that point in time if necessary.
